### PR TITLE
Add CI check to verify that benchmark queries return the expected results

### PR DESCRIPTION
**Pull request by andygrove**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/4016

# Rationale for this change

I would like to test that the benchmarks are producing the correct results so that we catch regressions. Some of the queries are producing very wrong results.

# What changes are included in this PR?

- Add `ci` feature to benchmark crate, which will verify results are correct
- Ignore failing tests
- Add new CI workflow
- Make benchmark crate usable as a dependency so that Ballista can use it (and avoid maintaining two copies of this code)

Output:

verify

```
running 22 tests
test tests::verify_q1 ... ok
test tests::verify_q10 ... ok
test tests::verify_q11 ... ignored
test tests::verify_q12 ... ok
test tests::verify_q13 ... ok
test tests::verify_q14 ... ignored
test tests::verify_q15 ... ok
test tests::verify_q16 ... ok
test tests::verify_q17 ... ignored
test tests::verify_q18 ... ok
test tests::verify_q19 ... ok
test tests::verify_q2 ... ok
test tests::verify_q20 ... ok
test tests::verify_q21 ... ok
test tests::verify_q22 ... ok
test tests::verify_q3 ... ok
test tests::verify_q4 ... ok
test tests::verify_q5 ... ok
test tests::verify_q6 ... ignored
test tests::verify_q7 ... ok
test tests::verify_q8 ... ok
test tests::verify_q9 ... ignored
```

# Are there any user-facing changes?

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -16,21 +16,24 @@
...
+members = ["datafusion/common", "datafusion/core", "datafusion/expr", "datafusion/jit", "datafusion/optimizer", "datafusion/physical-expr", "datafusion/proto", "datafusion/row", "datafusion/sql", "datafusion-examples", "benchmarks",
+]
 
 [profile.release]
 codegen-units = 1
 lto = true
+
+# the release profile takes a long time to build so we can use this profile during development to save time
+# cargo build --profile release-nonlto
+[profile.release-nonlto]
```
Comment (andygrove) [inline Cargo.toml:29]: The CI workflow uses this profile rather than `release` to try and save some time

File: benchmarks/src/bin/tpch.rs [hunk 2]
```diff
@@ -829,113 +618,140 @@ mod tests {
...
-    async fn q10() -> Result<()> {
+    async fn verify_q10() -> Result<()> {
         verify_query(10).await
     }
 
+    #[cfg(feature = "ci")]
+    #[ignore] // TODO produces wrong number of rows
     #[tokio::test]
-    async fn q11() -> Result<()> {
+    async fn verify_q11() -> Result<()> {
```
Comment (andygrove) [inline benchmarks/src/bin/tpch.rs:686]: fails due to https://github.com/apache/arrow-datafusion/issues/4023

File: benchmarks/src/bin/tpch.rs [hunk 3]
```diff
@@ -829,113 +618,140 @@ mod tests {
...
-    async fn q5() -> Result<()> {
+    async fn verify_q5() -> Result<()> {
         verify_query(5).await
     }
 
+    #[cfg(feature = "ci")]
+    #[ignore] // TODO produces wrong result
     #[tokio::test]
-    async fn q6() -> Result<()> {
+    async fn verify_q6() -> Result<()> {
```
Comment (andygrove) [inline benchmarks/src/bin/tpch.rs:654]: https://github.com/apache/arrow-datafusion/issues/4024

**Review by Dandandan — APPROVED** [review]
🚀

**Pull request merged**